gbus/bus.go

```diff
@@ -488,23 +489,67 @@ func (b *DefaultBus) sendWithTx(ctx context.Context, ambientTx *sql.Tx, toServic
 
 func (b *DefaultBus) returnDeadToQueue(ctx context.Context, ambientTx *sql.Tx, publishing *amqp.Publishing) error {
 	if !b.started {
-		return errors.New("bus not strated or already shutdown, make sure you call bus.Start() before sending messages")
+		return errors.New("bus not started or already shutdown, make sure you call bus.Start() before sending messages")
+	}
+
+	targetQueue, ok := publishing.Headers["x-first-death-queue"].(string)
+	if !ok {
+		return fmt.Errorf("bad x-first-death-queue field - %v", publishing.Headers["x-first-death-queue"])
+	}
+	exchange, ok := publishing.Headers["x-first-death-exchange"].(string)
+	if !ok {
+		return fmt.Errorf("bad x-first-death-exchange field - %v", publishing.Headers["x-first-death-exchange"])
+	}
+	routingKey, err := extractRoutingKey(publishing.Headers)
+	if err != nil {
+		return err
 	}
-	//publishing.Headers.
-	exchange := fmt.Sprintf("%v", publishing.Headers["x-first-death-exchange"])
-	routingKey := fmt.Sprintf("%v", publishing.Headers["x-first-death-queue"])
+
+	publishing.Headers[ResurrectedHeaderName] = true
+	publishing.Headers["x-first-death-routing-key"] = routingKey
+	// publishing.Headers["x-first-death-exchange"] is not deleted and kept as is
 
 	delete(publishing.Headers, "x-death")
 	delete(publishing.Headers, "x-first-death-queue")
 	delete(publishing.Headers, "x-first-death-reason")
-	delete(publishing.Headers, "x-first-death-exchange")
+
+	b.Log().
+		WithField("targetQueue", targetQueue).
+		WithField("firstDeathRoutingKey", routingKey).
+		WithField("firstDeathExchange", exchange).
+		Info("returning dead message to queue...")
 
 	send := func(tx *sql.Tx) error {
-		return b.publish(tx, exchange, routingKey, publishing)
+		// Publishing a "resurrected" message is done directly to the target queue using the default exchange
+		return b.publish(tx, "", targetQueue, publishing)
 	}
 	return b.withTx(send, ambientTx)
 }
 
+func extractRoutingKey(headers amqp.Table) (result string, err error) {
```

Comment:
I am wondering when would rabbitmq place multiple entries in the x-death list and if it can return a list then maybe we should support that as well.
WDYT ?

Reply:
yea I did wonder about this, I naively assumed the most recent death is the routing key I am searching for, and I think for now that assumption can work.

from what I understand if a message gets rejected multiple times after bouncing around and getting sent to dead letter queues, each death would add an entry to the x-death list with new entries always getting added to start of the list.
The "x-first-death-exchange", "x-first-death-queue", "x-first-death-reason" always refer to the first death and it would have been great had rabbit had a "x-first-death-routing-key" but it doesnt.

Reply:
I agree, from what I see in the rabbitmq doc then rabbitmq maintains the x-death list between reject cycles and always prepends the last reject details to the list so I think your assumption is correct